genesis/src/lockups.rs

```diff
@@ -0,0 +1,160 @@
+    cliff_epoch: Epoch,
+    /// number of post-cliff events
+    expiration_fraction: f64,
+    /// time between each post-cliff event
+    expiration_epochs: Epoch,
+}
+
+impl Lockups {
+    pub fn by_years(
+        cliff_fraction: f64,   // first cliff fraction
+        cliff_year: f64,       // first cliff time, in years
+        expirations: usize,    // number of follow-on expirations
+        expiration_years: f64, // years between each following expiration
+        epoch_schedule: &EpochSchedule,
+        tick_duration: &Duration,
+        ticks_per_slot: u64,
+    ) -> Self {
+        let cliff_slot = years_as_slots(cliff_year, tick_duration, ticks_per_slot) as u64;
+        let cliff_epoch = epoch_schedule.get_epoch(cliff_slot);
+
+        // assumes that the first cliff is after any epoch warmup
+        let first_expiration_slot =
+            years_as_slots(cliff_year + expiration_years, tick_duration, ticks_per_slot) as u64;
+        let expiration_epochs = epoch_schedule.get_epoch(first_expiration_slot) - cliff_epoch;
+
+        Self::by_epochs(cliff_fraction, cliff_epoch, expirations, expiration_epochs)
+    }
+
+    pub fn by_epochs(
+        cliff_fraction: f64,      // first cliff fraction
+        cliff_epoch: Epoch,       // first cliff epoch
+        expirations: usize,       //  number of follow-on expirations
+        expiration_epochs: Epoch, // epochs between each following expiration
+    ) -> Self {
+        let expiration_fraction = (1.0 - cliff_fraction) / expirations as f64;
+
+        Self {
+            prev_fraction: 0.0,
+            expiration: 0,
+            expirations,
+            cliff_fraction,
+            cliff_epoch,
+            expiration_fraction,
+            expiration_epochs,
+        }
+    }
+}
+
+impl Iterator for Lockups {
+    type Item = Expiration;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if 0 == self.expiration {
+            self.prev_fraction = self.cliff_fraction;
+            self.expiration += 1;
+
+            Some(Expiration {
+                prev_fraction: 0.0,
+                fraction: self.cliff_fraction,
+                epoch: self.cliff_epoch,
+            })
+        } else if self.expiration <= self.expirations {
+            // save these off
+            let (prev_fraction, expiration) = (self.prev_fraction, self.expiration);
+
+            // move forward, tortured-looking math comes from wanting to reach 1.0 by the last
+            //  expiration
+            self.prev_fraction =
+                1.0 - (self.expirations - expiration) as f64 * self.expiration_fraction;
+            self.expiration += 1;
+
+            Some(Expiration {
+                prev_fraction,
+                fraction: self.prev_fraction,
+                epoch: self.cliff_epoch + expiration as u64 * self.expiration_epochs,
+            })
+        } else {
+            None
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct Expiration {
```

Comment:
What's this thing?

Reply:
description of what happens at some point in time

Reply:
This answer needs to be in the code, not in a review comment.